### DISCUSSION TITLE
[chapter12] 커맨드/쿼리 분리 패턴

### DISCUSSION THREAD

**Discussion by coalong** [cowaine/is-my-code-that-strange#119] [Chapter 11-12 (주석, 메서드)] [answered]
p.266 에서 커맨드/쿼리 분리 내용이 나옵니다. gainAndGetPoint() 메서드를 gainPoint와 getPoint 메서드로 분리하는 것을 추천하는데 AtomicInteger라는 클래스가 있습니다. 여기는 동시성 제어를 위해 getAndIncrement, getAndDecrement 등등 커맨드/쿼리 분리가 되어있지 않은데, 이건 잘못된 설계일까요?

**Comment by firewoody237** [comment, marked as answer]
`AtomicInteger`는 Thread-Safe하게 구현된 특수한 Integer인데요,

Thread-Safe의 특성을 위해 위와 같은 메서드들이 존재하는 걸로 생각됩니다.

예를 들면
A Thread : Get() ------------> Update()
B Thread : --------Update()-----------

처럼 된다면, `AtomicInteger`는 Thread-Safe하다고 볼 수 없지 않을까요?

**Comment by corock** [comment]
잘 모르겠어서 GPT 에 물어보니 다음처럼 답변해 주네요.

> **CQRS와 Atomic 클래스 설계의 관계**
> **CQRS는 애플리케이션 아키텍처 레벨에서의 설계 원칙**입니다. 주로 도메인 모델이나 데이터 저장 계층에서 명령과 질의를 분리하는 데 사용됩니다.
> **Atomic 클래스는 낮은 레벨에서 동시성 제어를 하기 위한 도구**입니다. 이 클래스의 메서드는 단일 연산에서 상태를 변경하고 조회하는 동작을 안전하게 수행하기 위해 만들어졌습니다.
>
> 따라서 **Atomic가 커맨드/쿼리 분리를 따르지 않는 것은 설계의 맥락이 서로 다르기 때문**이지, 잘못된 설계라고 볼 수는 없습니다. Atomic 클래스의 주된 목적은 동시성을 효율적으로 관리하는 것이므로, 명령과 조회를 분리하지 않았다고 해서 이것이 잘못된 설계는 아닙니다.
>
> **결론**
> **CQRS와 Atomic 클래스는 서로 다른 맥락에서의 목적을 가진 설계 원칙입니다.** CQRS는 시스템 전반의 명령과 조회를 분리하여 성능과 유지보수성을 향상시키는 데 주요 목적이 있고, Atomic 클래스는 낮은 레벨의 동시성 문제를 해결하는 도구입니다.
> 따라서 Atomic 같은 클래스가 CQRS 원칙을 따르지 않는 것은 각 클래스의 설계 원칙과 목적이 다르기 때문이지, 잘못된 설계라고 볼 필요는 없습니다.
